File: frontend/ios/Hive/Hive/Location.swift
//
//  Location.swift
//  Hive
//
//  Created by Chuck Onwuzuruike on 12/30/18.
//  Copyright © 2018 Chuck Onwuzuruike. All rights reserved.
//

import Foundation
import CoreLocation

class Location {
    
    class Area : Equatable {
        public var latStr: String
        public var lonStr: String
        public var city: String
        public var state: String
        public var country: String
        
        init() {
            latStr = ""
            lonStr = ""
            city = ""
            state = ""
            country = ""
        }
        
        func toJson() -> [String: Any] {
            var json = [String: Any]()
            json["latitude"] = latStr
            json["longitude"] = lonStr
            json["city"] = city
            json["state"] = state
            json["country"] = country
            return json
        }
        
        func toString() -> String {
            return city + ", " + country
        }
        
        static func ==(lhs: Area, rhs: Area) -> Bool {
            return lhs.latStr == rhs.latStr && lhs.lonStr == rhs.lonStr
        }
    }
    
    public private(set) var location: CLLocation
    public var area: Area
    public var latStr: String
    public var lonStr: String
        
    init() {
        location = CLLocation(latitude: 0, longitude: 0)
        latStr = ""
        lonStr = ""
        area = Area()
    }

    init(locationJson: [String: Any]) {
        self.latStr = locationJson["latitude"] as! String
        self.lonStr = locationJson["longitude"] as! String
        self.area = Area()
        if locationJson["area"] != nil {
            let lJson = locationJson["area"] as! [String: Any]
            self.area.latStr = lJson["latitude"] as! String
            self.area.lonStr = lJson["longitude"] as! String
            self.area.city = lJson["city"] as! String
            self.area.state = lJson["state"] as! String
            self.area.country = lJson["country"] as! String
        }
        let lat: Double = Double(latStr) ?? 0
        let lon: Double = Double(lonStr) ?? 0
        self.location = CLLocation(latitude: lat, longitude: lon)
    }
    
    static func jsonToLocation(json: [String: Any]) -> Location {
        return Location(locationJson: json)
    }
    
    func toJson() -> [String: Any] {
        var json = [String: Any]()
        json["latitude"] = latStr
        json["longitude"] = lonStr
        json["area"] = area.toJson()
        return json
    }
}

File: frontend/android/app/src/main/java/coloredcoded/hive/client/Location.java
package coloredcoded.hive.client;

import org.json.simple.JSONObject;

import java.io.Serializable;

public class Location implements Serializable {
	
	public static class Area implements Serializable {
		private String latStr;
		private String lonStr;
		private String city;
		private String state;
		private String country;
		
		public Area() {
			latStr = "";
			lonStr = "";
			city = "";
			state = "";
			country = "";
		}
		
		public Area(String lat, String lon, String city, String state, String country) {
			latStr = lat;
			lonStr = lon;
			this.city = city;
			this.state = state;
			this.country = country;
		}
		
		public String getLatitude() {
			return latStr;
		}
		
		public String getLongitude() {
			return lonStr;
		}
		
		public String getCity() {
			return city;
		}
		
		public String getState() {
			return state;
		}
		
		public String getCountry() {
			return country;
		}
		
		public JSONObject toJSON() {
			JSONObject json = new JSONObject();
			json.put("latitude", latStr);
			json.put("longitude", lonStr);
			json.put("city", city);
			json.put("state", state);
			json.put("country", country);
			return json;
		}
	}
	
	/*
	 * This class represents a geographical location.
	 */

	private Area area;  // Always given to us by the server.
	private String latStr;
	private String lonStr;
	
	public Location(String lat, String lon) {
		area = new Area();
		latStr = lat;
		lonStr = lon;
	}

	public Location(String lat, String lon, Area area) {
		this.area = area;
		latStr = lat;
		lonStr = lon;
	}
	
	public Location() {
		area = new Area();
		latStr = "";
		lonStr = "";
	}
	
	public static Location jsonToLocation(JSONObject locationJson) {
		return new Location(locationJson);
	}
	
	private Location(JSONObject locationJson) {
		latStr = (String) locationJson.get("latitude");
		lonStr = (String) locationJson.get("longitude");
		if (locationJson.containsKey("area")) {
			JSONObject areaJson = (JSONObject) locationJson.get("area");
			area = new Area(
					(String) areaJson.get("latitude"),
					(String) areaJson.get("longitude"),
					(String) areaJson.get("city"),
					(String) areaJson.get("state"),
					(String) areaJson.get("country"));
		}
	}

	public Area getArea() {
		return area;
	}

	public String getLatitude() {
		return latStr;
	}

	public String getLongitude() {
		return lonStr;
	}

	public JSONObject toJSON() {
		JSONObject json = new JSONObject();
		json.put("latitude", latStr);
		json.put("longitude", lonStr);
		json.put("area", area.toJSON());
		return json;
	}
	
}
